Feature 3 overlapping shenanigans

Firebrandt opened the pull request:
This is a PR I'm making for the sake of getting at least partial feedback. Recommend reversion protocol be set up.
Adds a complete interactor, and a dialog window that interacts with it, but that does not satisfy the demands of the use case due to incomplete dependency for the moment. (And dire need of restructuring).

Known issues:
- Straight up doesn't connect to the timetable generator, since that doesn't exist yet. Without intermediate data the use case cannot be run properly. There's an ersatz kind of test case, (it at least RUNS, and the algorithm should work assuming the intermediate data from the timetable generator is entered at some point, but this is pretty bad). 
- Architecture is very not clean, but I will work on that! Notable issues include forgetting about input and output boundaries, which I will add in and restructure. Also dependency injection remains to be done, but will be handled.
- Connection to the main menu UI is incomplete.
- Connection to Hans' use case (another intermediate) is incomplete, but this is my fault for the moment.
- File structure is off, this was deferred.
- Tests incomplete, essentially none.
- Quite a few warnings. 

However, the algorithms are at least complete.
Sorry TA Hindawi, this is half baked kibble, but I'd still appreciate any feedback you can offer I guess.

Firebrandt committed:
Made the Course abstract entity.

Firebrandt committed:
Made the Course abstract entity.

michaelhsj committed:
dnc

Firebrandt committed:
Initial commit because a halfways workable dialog appears.

Firebrandt committed:
Created outline for TimeTableMatchInteractor.java and began implementing some of it.

Firebrandt committed:
Got a draft of the TimeTableMatchInteractor working.

Firebrandt committed:
Implemented java's Flow interfaces to (kind of) get bastardized communication between the controller and the dialog.

Firebrandt committed:
Restructuring architecture of program, which is currently broken. Made an outline of how it functions in the more propoer version of clean architecture.

Firebrandt committed:
Merge remote-tracking branch 'origin/feature-0-base-features' into feature-0-base-features

Firebrandt committed:
Wrote a kind of test case (sorta, mostly just for me though it could be turned into a test case later). Architecture substantially fixed -- repairs still being made, though.

michaelhsj commented:
Hey Michael! It looks like you've got some code that can't quite yet compile, so I hesitate to approve a merge until it can do that.

You've already mentioned it, but I figured this is a good place to talk about what sorts of interfaces you'll need going forward. Your TimeTableMatchCalculateSectionHoursInteractor should implement an InputBoundary of some sort, most likely with the method determineBestMatchingTimetable. Your dialog presenter should implement an OutputBoundary, and again the interface should have all functions that your interactor calls of the presenter.

I can only guess what InputDialogMakingInteractor is, but my advice is that unless this is an entirely separate use case, it might be a task that the presenter does instead? If this has to do with alerting the user and presentation logic, it might be better suited in the presenter. And lastly, whatever View you end up implementing can also implement a View Model, which does something similar.

Also, I noticed that as it is, your Controller seems to be taking Timetable entities as an input when ideally it should only be taking inputs from your View. The only part of your use case that should really be holding or passing around entities is your interactor, so I would suggest using either the TimetableModel that's going to be merged into main from branch 9, or making your interactor an observer to keep track of the timetables instead. Also, I would consider wrapping up the arguments you give the interactor as a RequestModel, but that might not be entirely necessary depending on how much smaller your argument list becomes. 

Looking at the TimeTableMatchCalculateSectionHoursInteractor, a part of me wonders if there's a way to extract a method from calculateTimetableOverlap, if only to make the code more readable.

Also, many of your input and output boundary interfaces have the public access modifier, which is unnecessary and raises IntelliJ warnings, so make sure to clean those up before merging.

You've got some good code so far, so keep working at it! Cleaning up the architecture is the hardest part- everything else will come quite easily.

Firebrandt committed:
Merge remote-tracking branch 'origin/main'

Firebrandt committed:
Merge branch 'main' into feature-3-overlapping-shenanigans

Firebrandt committed:
Pulled in a bunch of stuff from main() that people have merged previously so that I can integrate.

(At least, partially) cleaned up my architecture to fit Clean Architecture guidelines and to use Request and Response models. This should help break dependency up a bit.

Clean warnings in later part.

Firebrandt committed:
Laid the groundwork to connect to Hans' and JD's use cases :).

Firebrandt committed:
Wrote some simple test cases for CalculateSectionHoursInteractor.

Firebrandt committed:
Finished some tests for the TimeTableMatchInteractor.

Firebrandt committed:
Slightly restructured my use case files.

Firebrandt committed:
Um, for some reason this wasn't caught in the last commit. Strange.

Firebrandt committed:
Did some additional work and restructuring. Chiefly, sets up more to integrate with JD's use case, and edits the InputDialog to support passing additional data relating to the setup of the OverlapMaximization Use Case. Sets OverlapMaximizationController and TimeTableMatchInteractor up such that the other parts of the use case can listen for data from the InputDialog on things like the selected timetable.

Essentially, the skeleton is all done and I should hopefully just need to plug in JD's stuff now :) for intermediate input. And maybe link it to Kai's and add unit tests or something.

moonsdust committed:
Added TimetableGatewayTest

moonsdust committed:
Merge remote-tracking branch 'origin/feature-5-breadth-requirement-recommendation-add-test' into feature-6-file-io

moonsdust committed:
refactored files and added commenst to JSON file

moonsdust committed:
refactored files and added comment to JSON file

Kai15116 committed:
Merge branch 'feature-5-breadth-requirement-recommendation-clean' of https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name into feature-9-finish-main-menu-ui

 Conflicts:
	src/main/java/screens/MainUI.java

Kai15116 committed:
Resolved issue with pulling from other branch

moonsdust committed:
Implemented Publisher in SessionGatewayInteractor

moonsdust committed:
Merge remote-tracking branch 'origin/feature-9-finish-main-menu-ui' into feature-6-file-io

Kai15116 committed:
Organized variable name

moonsdust committed:
Made SessionFileController not return anything and fixed java flow implementation of publisher

Kai15116 committed:
Merge branch 'feature-3-overlapping-shenanigans' of https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name into feature-10-finish-timetable-view

Firebrandt committed:
Did some additional work and restructuring. Chiefly, sets up more to integrate with JD's use case, and edits the InputDialog to support passing additional data relating to the setup of the OverlapMaximization Use Case. Sets OverlapMaximizationController and TimeTableMatchInteractor up such that the other parts of the use case can listen for data from the InputDialog on things like the selected timetable.

Essentially, the skeleton is all done and I should hopefully just need to plug in JD's stuff now :) for intermediate input. And maybe link it to Kai's and add unit tests or something.

-- Kind of another commit, forgot to push that yesterday.

Also moved Flow.Subscriber() to the OverlapMaximizationController from team feedback. Also refined and completed the implementation of that a little bit more. Refactored TimeTableMatchInteractor accordingly so that it doesn't have to listen to the Dialog (a teammate told me that was a bad idea). Instead the Controller will listen to the Dialog, and then use that to call the Interactor's methods.

Interactor no longer stores mainTable, and timetables - instead, those are method parameters, because it doesn't make sense to make a new interactor / update existing one's fields for every new pair. This way we can make it a singleton!

Kai15116 committed:
Merge branch 'feature-3-overlapping-shenanigans' of https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name into feature-10-finish-timetable-view

Kai15116 committed:
Resolved (not fully) merge conflict

Kai15116 committed:
Merge branch '7-feature-7-blacklistwhitelist-domenttest' of https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name into feature-9-finish-main-menu-ui

Kai15116 committed:
Modified to resolve conflict

Kai15116 committed:
Merge branch 'feature-13-finishing-sorter' of https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name into feature-9-finish-main-menu-ui

 Conflicts:
	src/main/java/screens/MainUI.java

Kai15116 committed:
Merge branch 'feature-5-breadth-requirement-recommendation-clean' of https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name into feature-9-finish-main-menu-ui

Kai15116 committed:
Merge branch 'feature-10-finish-timetable-view' of https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name into feature-9-finish-main-menu-ui

 Conflicts:
	src/main/java/blacklist_whitelist_use_case/frameworks_and_drivers/ConstraintsInputScreen.java

Kai15116 committed:
Modified to resolve conflict

Kai15116 committed:
Merge branch 'feature-13-finishing-sorter' of https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name into feature-9-finish-main-menu-ui

Kai15116 committed:
Merge branch 'feature-6-file-io' of https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name into feature-9-finish-main-menu-ui

Kai15116 committed:
Merge branch 'feature-12-finish-timetable-editing' of https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name into feature-9-finish-main-menu-ui

 Conflicts:
	src/main/java/RecommendBRTestMain.java
	src/main/java/edit_timetable_use_case/AddCourseInteractor.java
	src/main/java/edit_timetable_use_case/application_business/RemoveCourseInteractor.java
	src/main/java/screens/EditTimetableScreen.java
	src/main/java/screens/MainUI.java

Kai15116 committed:
Organized main in MainUI

Kai15116 committed:
Merge branch 'feature-12-finish-timetable-editing' of https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name into feature-9-finish-main-menu-ui

Kai15116 committed:
Merge branch 'feature-11-cleaning-up-retrieve-timetable' of https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name into feature-9-finish-main-menu-ui

Kai15116 committed:
Added comment

YahyaElgabra committed:
implemented TimetablesSortInteractor, documented my code

next steps: integrate with mainmenu and timetableUI, flow, tests

YahyaElgabra committed:
Merge remote-tracking branch 'origin/feature-13-finishing-sorter' into feature-13-finishing-sorter

YahyaElgabra committed:
checking if github still hates enhanced for loops

Kai15116 committed:
Integrated overlap maximization to TimetableUI

Kai15116 committed:
Changed MainUI so it shows the edit and display buttons only if user import timetable file

Kai15116 committed:
Merge branch 'feature-12-finish-timetable-editing' of https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name into feature-9-finish-main-menu-ui

YahyaElgabra committed:
Implemented flow publisher, still didn't integrate it with the rest of my use case

YahyaElgabra committed:
Merge branch 'main' into feature-13-finishing-sorter

# Conflicts:
#	src/main/java/screens/MainUI.java

YahyaElgabra committed:
no implementations yet, stuck on how to convert TimetableViewModel into TimetableModel

YahyaElgabra committed:
Merge branch 'feature-9-finish-main-menu-ui' into feature-13-finishing-sorter

# Conflicts:
#	src/main/java/blacklist_whitelist_use_case/frameworks_and_drivers/ConstraintsInputScreen.java
#	src/main/java/screens/MainUI.java

YahyaElgabra committed:
I will go to WAR with swing

WHY CANT IT JUST LET ME SWITCH SCREENS AAAAAAAAAAAAAAH

YahyaElgabra committed:
started doing tests but stopped cuz i need sleep
for mainmenu integration: ask michael about swing
for ttui integration: ask people about converting from timetableviewmodel down
for flow: done

Firebrandt committed:
DESIGN DECISION MADE:
- UNDID THE OBSERVER PATTERN SINCE I WAS RUNNING INTO A KIND OF OVERCOMPLICATING IT CODE SMELL. REFACTORGURU MENTIONS IT, BUT BASICALLY, THE PATTERN OVERCOMPLICATES THINGS AND ACTUALLY MAKES IT HARDER.

- More refactoring since I found out some clean architecture was wrong. Turns out I need to pass in ViewModel to the InputDialog, so I've made those too.

- Added some skeletal support for TimetableUI from Kai. Hopefully this is good!

Firebrandt committed:
Factored out another CA violation into yet another convert! !!!
!!! !!! :D

YahyaElgabra committed:
still implementing AllTimetablesController

YahyaElgabra committed:
still implementing AllTimetablesController

YahyaElgabra committed:
observer pattern stuff

Firebrandt committed:
Merge remote-tracking branch 'origin/8-feature-2-generating-with-course-list' into feature-3-overlapping-shenanigans

# Conflicts:
#	src/main/java/blacklist_whitelist_use_case/frameworks_and_drivers/ConstraintsInputScreen.java

Firebrandt committed:
Merge remote-tracking branch 'origin/feature-13-finishing-sorter' into feature-3-overlapping-shenanigans

# Conflicts:
#	build.gradle
#	src/main/java/display_timetable_use_case/interface_adapters/TimetableUI.java
#	src/main/java/display_timetable_use_case/interface_adapters/TimetableViewBlockModel.java
#	src/main/java/screens/MainUI.java

JD-Grimm committed:
Modified addTimetableCourse to follow clean architecture better with Kai's help

JD-Grimm committed:
Modified generateTimetable to properly generate a timetable list

Firebrandt committed:
Wrote the bones of um, a way to get data by listening to JD. Also happened to grab some stuff from Yahya - thx to him for implementing that publisher <3.

Firebrandt committed:
Wrote the bones of um, a way to get data by listening to JD. Also happened to grab some stuff from Yahya - thx to him for implementing that publisher <3.

Firebrandt committed:
Made a kind of mechanism to wait for JD's thing to finish. Ready to try running it now... I guess?

Firebrandt committed:
Set up a draft for my initialization code.

Firebrandt committed:
Fixed a few issues with the draft initialization.

Firebrandt committed:
Fixed a few issues with the draft initialization.

Firebrandt committed:
Fixed a few issues with the draft initialization.

Firebrandt committed:
Commit just to save

Firebrandt committed:
Merge remote-tracking branch 'origin/8-feature-2-generating-with-course-list' into feature-3-overlapping-shenanigans

Firebrandt committed:
Fixed architecture problem with the Controller (output goes through a Presenter to the View now :) )

Firebrandt committed:
Fixed unit tests for one of the Converters.

Firebrandt committed:
Made unit tests for the other converter.

Firebrandt committed:
Fixed my interactor unit tests >:D

Firebrandt committed:
Made new unit tests. Doing super integration now.

Refactored to not kill the main frame (instead refactors it).

Firebrandt committed:
Fixed my file layout.

Firebrandt committed:
Added a new function to update UI to the InputDialog.

Firebrandt committed:
Refined the UI resetting stuff a little bit.

michaelhsj reviewed (commented):
Responding to the changes made here: https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name/pull/86#issue-1484118368

Okay! So this is overall looking much, much cleaner than before, so I'm loving the architecture. I've left some comments on the code that you can take a look at, but the long and short is as follows:

I would change the overlap package name to something more descriptive and technical. Right now it seems quite loose.

Your OverlapGeneratedTimetableRelayInteractor doesn't seem to implement any input boundaries. I'm assuming it's the case that only other interactors call it, in which case I don't think there's any problem. But if a controller ends up calling it, make sure it implements some InputBoundary.

Your TimeTableMatchInteractor has some good code, but calculateTimetableOverlap has a 5 levels of nesting, which is a little smelly. If we have some time in the future, we should plan on extracting some of the method to make it easier to read and extend. It's a priority, but the lowest on the list.

TimetableMatchInputBoundary could use a couple sentences of documentation.

The tests in TestCalculateSectionHoursInteractor are all commented out right now. Do these tests work?

Please scan over all of your code for any yellow warnings on IntelliJ. There are too many to list, but I've pointed out a few and  they need to be cleared out before we merge to main.

(For the TA: many of the code comments that are mentioned in this review are here: https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name/pull/86#issue-1484118368
 sorry for all the branches!)

michaelhsj commented on src/main/java/overlap_crap_fix_locations_later/application_business/OverlapOutputBoundary.java:
```diff
@@ -0,0 +1,20 @@
+    /**
+     * Pass viewModels to the dialog, given a list of TimetableModels
+     **/
+    public void passViewModelsToDialog(List<TimetableModel> timetableModels);
```
The public access modifier here is unnecessary for an interface. Make sure to remove this (and any other instances in your interfaces).

Firebrandt replied:
Killed it!

Firebrandt replied:
Removed the access modifier.

michaelhsj commented on src/main/java/overlap_crap_fix_locations_later/application_business/SectionHoursInputBoundary.java:
```diff
@@ -0,0 +1,12 @@
+public interface SectionHoursInputBoundary {
+
+    /** calculate the length, in Hours, of a section. **/
+    public Double calculateHoursOfSection(SectionModel section);
```
Same here. No need for the public modifier here.

michaelhsj commented on src/main/java/overlap_crap_fix_locations_later/application_business/TimeTableMatchInteractor.java:
```diff
@@ -0,0 +1,105 @@
+                            // If they're in the same section in the same course, the two people can go together :).
+                            // Hey, what if we used a companion object to not have to make a new one each time? Or just made this into an object.
+
+                            ArrayList<BlockModel> blockModels = new ArrayList<>();
```
It looks like this variable is completely unused, which raises a warning on IntelliJ.

Firebrandt replied:
No longer use it!

michaelhsj commented on src/main/java/overlap_crap_fix_locations_later/interface_adapters/ModelToOverlapViewModelConverter.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Convenience method to turn a BlockModel into an OverlapTimetableBlockViewModel.
+     *
+     * @param blockModel
```
blockModel is missing a description here.

Firebrandt replied:
The new version fixes this :p.

Firebrandt committed:
Implement Michael H's revisions. Fixed warnings in my files.

Added a strings.properties file to store UI strings. Centralizing them in one place could be a good idea to increase UI maintainability (what if we need to translate this abominable application into a different language?)

Firebrandt commented:
Ok! So I made a kind of duplicate pull request, my bad. 
For the changes in this one, see https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name/pull/86#issue-1484118368 - the comment michael H alluded to. 

Made an additional design decision just now to put UI strings in, strings.properties, a central file. Keeping all our UI strings in one place should make the UI portion of the code more maintainable; a central file makes it easier to proofread and keep track of the UI strings in our app, and if we want to re-use a string resource, that becomes easier too (saves us the trouble of retyping the literal).

In the event that we translate this application to another language, having them all in one places makes translation easier too. 

I added a commit implementing Michael H's revisions, and fixing the warnings in my use case's files.

michaelhsj reviewed (approved):
Looks good! Make sure to resolve your merge conflicts before merging. My advice would be to merge main into your branch and resolve it in IntelliJ first, but you can also get it done using Github's merge UI.
